pkg/ebpf/uprobes/attacher.go

```diff
@@ -0,0 +1,901 @@
+// to shared libraries or executables, this structure tells the attacher which ones to
+// select and to which targets to do it.
+type AttachRule struct {
+	// LibraryNameRegex defines which libraries should be matched by this rule
+	LibraryNameRegex *regexp.Regexp
+	// ExecutableFilter is a function that receives the path of the executable and returns true if it should be matched
+	ExecutableFilter func(string, *ProcInfo) bool
+	// Targets defines the targets to which we should attach the probes, shared libraries and/or executables
+	Targets AttachTarget
+	// ProbesSelectors defines which probes should be attached and how should we validate
+	// the attachment (e.g., whether we need all probes active or just one of them, or in a best-effort basis)
+	ProbesSelector []manager.ProbesSelector
+	// ProbeOptionsOverride allows the user to override the options for a probe that are inferred from the name
+	// of the probe. This way the user can set options such as manual return detection or symbol names for probes
+	// whose names aren't valid C identifiers.
+	ProbeOptionsOverride map[string]ProbeOptions
+}
+
+// canTarget returns true if the rule matches the given AttachTarget
+func (r *AttachRule) canTarget(target AttachTarget) bool {
+	return r.Targets&target != 0
+}
+
+func (r *AttachRule) matchesLibrary(path string) bool {
+	return r.canTarget(AttachToSharedLibraries) && r.LibraryNameRegex != nil && r.LibraryNameRegex.MatchString(path)
+}
+
+func (r *AttachRule) matchesExecutable(path string, procInfo *ProcInfo) bool {
+	return r.canTarget(AttachToExecutable) && (r.ExecutableFilter == nil || r.ExecutableFilter(path, procInfo))
+}
+
+// getProbeOptions returns the options for a given probe, checking if we have specific overrides
+// in this rule and, if not, using the options inferred from the probe name.
+func (r *AttachRule) getProbeOptions(probeID manager.ProbeIdentificationPair) (ProbeOptions, error) {
+	if r.ProbeOptionsOverride != nil {
+		if options, ok := r.ProbeOptionsOverride[probeID.EBPFFuncName]; ok {
+			return options, nil
+		}
+	}
+
+	symbol, isManualReturn, err := parseSymbolFromEBPFProbeName(probeID.EBPFFuncName)
+	if err != nil {
+		return ProbeOptions{}, err
+	}
+
+	return ProbeOptions{
+		Symbol:         symbol,
+		IsManualReturn: isManualReturn,
+	}, nil
+}
+
+// Validate checks whether the rule is valid, returns nil if it is, an error message otherwise
+func (r *AttachRule) Validate() error {
+	var errs []string
+
+	if r.Targets == 0 {
+		errs = append(errs, "no targets specified")
+	}
+
+	if r.canTarget(AttachToSharedLibraries) && r.LibraryNameRegex == nil {
+		errs = append(errs, "no library name regex specified")
+	}
+
+	for _, selector := range r.ProbesSelector {
+		for _, probeID := range selector.GetProbesIdentificationPairList() {
+			_, err := r.getProbeOptions(probeID)
+			if err != nil {
+				errs = append(errs, fmt.Sprintf("invalid probe name %s: %s", probeID.EBPFFuncName, err))
```

Comment:
I fill like `fmt.Sprintf("invalid probe name %s: %s", probeID.EBPFFuncName, err)` should be coming from `getProbeOptions`
In the future we can change the logic of `getProbeOptions` to have more errors that does not necessarily indicates a problem in the name of the probe

Reply:
Changed to a generic error message.